app/controllers/data_explorer/query_controller.rb

```diff
@@ -0,0 +1,236 @@
+  end
+
+  def set_group
+    @group = Group.find_by(name: params["group_name"])
+  end
+
+  def set_query
+    @query = DataExplorer::Query.find(params[:id])
+  end
+
+  def index
+    queries = DataExplorer::Query.where(hidden: false).order(:last_run_at, :name).to_a
+
+    database_queries_ids = DataExplorer::Query.pluck(:id)
+    Queries.default.each do |params|
+      attributes = params.last
+      query = DataExplorer::Query.new
+      query.id = attributes["id"]
+      query.sql = attributes["sql"]
+      query.name = attributes["name"]
+      query.description = attributes["description"]
+      query.user_id = Discourse::SYSTEM_USER_ID.to_s
+      queries << query unless database_queries_ids.include?(query.id)
+    end
+
+    render_serialized queries, DataExplorer::QuerySerializer, root: 'queries'
+  end
+
+  skip_before_action :check_xhr, only: [:show]
+  def show
+    check_xhr unless params[:export]
+
+    query = DataExplorer::Query.find(params[:id])
+
+    if params[:export]
+      response.headers['Content-Disposition'] = "attachment; filename=#{query.slug}.dcquery.json"
+      response.sending_file = true
+    end
+
+    return raise Discourse::NotFound if !guardian.user_can_access_query?(query) || query.hidden
+    render_serialized query, DataExplorer::QuerySerializer, root: 'query'
+  end
+
+  def groups
+    render_serialized(Group.all, BasicGroupSerializer)
+  end
+
+  def group_reports_index
+    return raise Discourse::NotFound unless guardian.user_is_a_member_of_group?(group)
+
+    respond_to do |format|
+      format.html { render 'groups/show' }
+      format.json do
+        queries = DataExplorer::Query
+          .where(hidden: false)
+          .joins("INNER JOIN data_explorer_query_groups data_explorer_query_groups 
+                    ON data_explorer_query_groups.query_id = data_explorer_queries.id 
+                    AND data_explorer_query_groups.group_id = #{group.id}")
+        render_serialized(queries, DataExplorer::QuerySerializer, root: 'queries')
+      end
+    end
+  end
+
+  def group_reports_show
+    return raise Discourse::NotFound if !guardian.group_and_user_can_access_query?(group, query) || query.hidden
+
+    respond_to do |format|
+      format.html { render 'groups/show' }
+      format.json do
+        render_serialized query, DataExplorer::QuerySerializer, root: 'query'
+      end
+    end
+  end
+
+  skip_before_action :check_xhr, only: [:group_reports_run]
+  def group_reports_run
+    return raise Discourse::NotFound if !guardian.group_and_user_can_access_query?(group, query) || query.hidden
+
+    run
+  end
+
+  def create
+    query = DataExplorer::Query.create!(params.require(:query).permit(:name, :description, :sql).merge(user_id: current_user.id, last_run_at: Time.now).permit!)
+    group_ids = params.require(:query)[:group_ids]
+    group_ids&.each do |group_id|
+      query.query_groups.find_or_create_by!(group_id: group_id)
+    end
```

Comment:
We should execute all the queries in a single transaction so that one query failing does not leave us in a weird state.